Catch `Exception` and not `BaseException` in the `Connection`

kristjanvalur opened the pull request:
### Pull Request check-list

_Please make sure to review and check all of these items:_

- [x] Does `$ tox` pass with this change (including linting)?
- [x] Do the CI tests pass with this change (enable it first in your forked repo and wait for the github action build to finish)?
- [x] Is the new or changed code fully tested?
- [ ] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [ ] Is there an example added to the examples folder (if applicable)?
- [x] Was the change added to CHANGES file?

### Description of change

Fixes Issue #2103:
Raising a BaseException inside a `Connection`'s send or receive oprations causes a `disconnect`().  BaseExceptions are commonly used for _out of band_ messages in Python and normally not be expecially acted upon, only be handled at the top-level 

See also:  #2356 which focuses on a different aspcect of error handling.

WisdomPill reviewed (commented):
LGTM!

Can you add some tests for it?

chayim commented:
@kristjanvalur Why did you catch Exception rather that BaseException? Was there an inheritance issue you saw?

WisdomPill commented:
@chayim have a look here #2103

kristjanvalur commented:
I can add tests.  As I explained in the defect, `BaseException` is usually not caught except in top level handlers, unless it is to implement something like `async_timeout.Timeout()`  .  `BaseException` was added to Python precisely to be a sort of _meta_ exception that regular exception handling doesn't touch, the `Exception` being the most general class usually touched in code that isn't trying to be too clever.

kristjanvalur commented:
In fact, I don't see why there is a `except Exception: self.disconnect(); raise` at all. Doesn't seem like something the Connection should do if it gets an _unexpected_ exception.  IMHO, this clause should be removed alltogether.

kristjanvalur commented:
Hm, now it is mysteriously failing.  it asserts on the "assert  pubsub.connection.is_connected"
I´m running tests both on windows and in a linux dev container using
`pytest tests/test_async/test_pubsub.py`.   I don't get the test failures above.

kristjanvalur commented:
Well, it seems I cracked the unittest problem I was having.  Turned out that only in Python 3.8 did asyncio.CancelledError get promoted to BaseException.
Fixed up the test to skip older python versions and added Mock-based tests for exception propagation for both async and blocking redis.

kristjanvalur commented:
any responses?  @WisdomPill ?

kristjanvalur commented:
I'll clean up the commit history a bit...

kristjanvalur commented:
yet another fluke test failure.

kristjanvalur commented:
I added the Connection.disconnect(nowait) optional arg.  When `true`, the disconnect() call will not attempt to wait for the flushing of the write buffer during close().  This is an async operation, and when calling "disconnect" as a response to an error, the async close can cause other errors to be raised.  Notably, if running within a "timeout" context, a `CancelledError` may get thrown.

In the case of an error, there should be no need to wait for a flush of the write buffer.

akx commented:
I bumped into this too – trying to e.g. ctrl+c out of py.test requires a couple of tries because `KeyboardInterrupt` (a `BaseException`) is caught where it really must not be.

kristjanvalur commented:
woo-hoo!

dvora-h commented:
Thanks for this! And sorry this (and your other PR's) waited so long...
